graphite_beacon/handlers/telegram.py

```diff
@@ -23,89 +29,149 @@ def init_handler(self):
         self.bot_ident = self.options.get('bot_ident')
         assert self.bot_ident, 'Telegram bot ident token is not defined.'
 
+        self.chatfile = self.options.get('chatfile')
+        if not self.chatfile: LOGGER.warning(NO_CHATFILE)
+
         self.client = httpclient.AsyncHTTPClient()
-        self.url = "https://api.telegram.org/bot%s/" % (self.token)
+        self.url = 'https://api.telegram.org/bot%s/' % (self.token)
 
-        self._chats = []
+        self._chats = self._get_chatlist()
         self._listen_commands()
 
-    def get_message(self, level, alert, value, target=None, ntype=None, rule=None):
-
-        msg_type = 'telegram' if ntype == 'graphite' else 'short'
-        tmpl = TEMPLATES[ntype][msg_type]
-        return tmpl.generate(
-            level=level, reactor=self.reactor, alert=alert, value=value, target=target).strip()
-
     @gen.coroutine
     def _listen_commands(self):
 
         self._last_update = None
 
         update_body = {"timeout": 2}
 
-        update_headers = {"Content-Type": "application/json"}
-
         while True:
+
             if self._last_update:
                 update_body.update({"offset": self._last_update + 1})
-            update_resp = self.client.fetch(
-                self.url + "getUpdates", headers=update_headers,
-                body=json.dumps(update_body), method="POST")
+
+            req = self._prepare_request('getUpdates', update_body)
+            update_resp = self.client.fetch(**req)
+
             update_resp.add_done_callback(self._respond_commands)
             yield gen.sleep(5)
 
     @gen.coroutine
     def _respond_commands(self, update_response):
+        chatfile = self.chatfile
 
         if update_response.exception():
             LOGGER.error(str(update_response.exception()))
 
-        update_content = update_response.result().body
-        if not update_content:
+        upd = update_response.result().body
+        if not upd:
             return
-        else:
-            update_content = json.loads(update_content)
 
-        for update in update_content["result"]:
-            if not update["message"].get('text'):
+        update_content = json.loads(upd.decode())
+        for update in update_content['result']:
+
+            update_id = update.get('update_id')
+            if not update_id: continue
+
+            message = update.get('message')  # fix key error here
+            if not message: continue
+
+            text = message.get('text')
+            if not text: continue
+
+            message_id = message.get('message_id')
+            if not message_id: continue
+
+            chat_id = message.get('chat', {}).get('id')
+            if not chat_id: continue
+
+            if not self._correct_activation(text, chat_id):
```

Comment:
> Also is this an exceptional circumstance worth logging?

Well, the absence of `message` field in `update` is normal (and not worth logging IMO) -- it happens when someone edits message in chat for example.

`_correct_activation` method is a whole different story, generally also not worth logging. Imagine chat conference (aka telegram group) with several participants including our bot, we would be logging every time someone posts. But we might consider logging when user tries to activate bot with wrong bot_ident. If someone thinks this case should be logged ill do it (should it be Logger.debug/Logger.warning/Logger.error?).

The rest of cases are not so clear to me, not sure whether all those situations are even possible. With exception of `if not update_id: continue` all those continue guards already were in code, i just put them all in one place one after another.
Will look closer into Telegram API, but im quite sure every update has update_id, so at least this `continue` is not needed.